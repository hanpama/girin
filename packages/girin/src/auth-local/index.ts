import * as argon2 from 'argon2';
import * as jwt from 'jsonwebtoken';

import { Model, field } from '../mongodb';
import { ObjectID } from 'bson';
import AuthLocalModule from './module';
import { globalEnvironment } from '@girin/environment';


function getPasswordSalt() {
  return globalEnvironment.get(AuthLocalModule).PASSWORD_SALT;
}
function getJWTSecretKey() {
  return globalEnvironment.get(AuthLocalModule).JWT_SECRET_KEY;
}

export type BaseUserClass<T extends BaseUser> = typeof BaseUser & {
  new(source: any): T;
}

export class BaseUser extends Model {

  @field()
  protected hashedPassword: string;

  public static async createUserWithPassword(source: { [fieldName: string]: any }, password: string) {

    const hashedPassword = await argon2.hash(password + getPasswordSalt());
    const user = await this.insert({ hashedPassword, ...source });
    return user;
  }

  public authenticate(password: string) {
    return argon2.verify(this.hashedPassword, password + getPasswordSalt());
  }

  public async changePassword(newPassword: string) {
    const hashedPassword = await argon2.hash(newPassword + getPasswordSalt());
    this.hashedPassword = hashedPassword;
    await this.$update({ $set: { hashedPassword } });

    return this;
  }

  public async encodeToken(password: string) {
    const authenticated = await this.authenticate(password);
    if (!authenticated) {
      throw new Error('Authentication Error');
    }
    const { hashedPassword, ...rest } = this.$source;
    return jwt.sign(rest, getJWTSecretKey());
  }

  public static decodeToken<T extends BaseUser>(this: BaseUserClass<T>, token: string): T {
    const result: any = jwt.verify(token, getJWTSecretKey());
    console.log(result);
    const user = new this(result);
    user._id = new ObjectID(result._id);
    return user;
  }
}
