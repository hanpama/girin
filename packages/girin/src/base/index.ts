export * from './scalar';
export * from './Definition';
export * from './MetadataStorage';
export * from './TypeExpression';
export * from './structure';
