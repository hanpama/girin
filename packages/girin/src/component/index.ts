export * from './Component';
export * from './decorators';
