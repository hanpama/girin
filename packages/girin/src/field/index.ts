export * from './Field';
export * from './InputField';
// export * from './mount';
