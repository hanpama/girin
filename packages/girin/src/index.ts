export { StringScalar, BooleanScalar, FloatScalar, IntScalar, IDScalar } from './builtins/scalar';
export { List, NonNull } from './type-expression/structure';
export * from './base';
export * from './metadata';
export * from './sdl';
export * from './getGraphQLType';
export * from './field';
export * from './type-expression';
