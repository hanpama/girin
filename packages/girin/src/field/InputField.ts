import { GraphQLInputFieldConfig, GraphQLInputType } from "graphql";
import { TypeExpression, TypeArg } from "../type-expression/TypeExpression";
import { MetadataStorage } from "../base/MetadataStorage";


export interface InputFieldProps {
  defaultValue?: any;
  description?: string;
  directives?: any;
}

export interface InputFieldReference {
  name: string;
  field: InputField;
  props: InputFieldProps;
}

export class InputField {

  protected input: TypeExpression;

  constructor(input?: TypeArg | TypeExpression) {
    if (input) {
      this.input = input instanceof TypeExpression ? input : new TypeExpression(input);
    }
  }

  public buildConfig(storage: MetadataStorage): GraphQLInputFieldConfig {
    const { input } = this;
    return {
      type: input.buildTypeInstance(storage) as GraphQLInputType,
    };
  }
}
