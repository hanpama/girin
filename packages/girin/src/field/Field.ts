import { GraphQLFieldConfigArgumentMap, GraphQLFieldConfig, GraphQLOutputType, GraphQLFieldResolver } from "graphql";
import { TypeExpression, TypeArg } from "../type-expression/TypeExpression";
import { MetadataStorage } from "../base/MetadataStorage";
import { InputFieldReference } from "./InputField";


export interface FieldProps {
  description?: string;
  deprecationReason?: string;
  directives?: any;
  resolve?: Function;
}

export interface FieldReference {
  name: string;
  field: Field;
  props: FieldProps;
}

export class Field {

  protected output: TypeExpression;
  protected args: InputFieldReference[];

  constructor(
    output?: TypeArg | TypeExpression,
    args?: InputFieldReference[],
  ) {
    if (output) {
      this.output = output instanceof TypeExpression ? output : new TypeExpression(output);
    }
    this.args = args || [];;
  }


  public buildResolver(storage: MetadataStorage): GraphQLFieldResolver<any, any> | undefined  {
    return undefined;
  }

  public buildArgs(storage: MetadataStorage): GraphQLFieldConfigArgumentMap {
    const { args } = this;
    return args.reduce((args, ref) => {
      args[ref.name] = Object.assign(ref.field.buildConfig(storage), ref.props);
      return args;
    }, {} as GraphQLFieldConfigArgumentMap);
  }

  public buildConfig(storage: MetadataStorage): GraphQLFieldConfig<any, any> {
    const { output } = this;

    return {
      type: output.buildTypeInstance(storage) as GraphQLOutputType,
      args: this.buildArgs(storage),
      resolve: this.buildResolver(storage),
    };
  }
}
