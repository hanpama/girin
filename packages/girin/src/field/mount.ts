import { Field, FieldReference, FieldProps } from "./Field";
import { MetadataStorage } from "../base/MetadataStorage";
import { globalMetadataStorage } from "../globalMetadataStorage";


export function mount(field: Field, props: FieldProps = {}, storage: MetadataStorage = globalMetadataStorage) {
  return (target: object, propertyKey: string) => {
    const reference: FieldReference = { field, name: propertyKey, props };
    globalMetadataStorage.registerFieldReference(reference, (target as any).constructor);
  }
}
