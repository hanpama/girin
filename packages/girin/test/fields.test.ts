import { defineType, gql, getGraphQLType, List } from "../src";
import { Field } from "../src/field/Field";
import { mount } from "../src/field/mount";
import { GraphQLObjectType, GraphQLSchema, graphql, printSchema } from "graphql";
import { InputField } from "../src/field/InputField";


@defineType(gql`
  type Room {
    name: String!
  }
`)
class Room {
  name: string;

  @mount(new Field(List.of('String')))
  tags: string[];
}

class RoomsField extends Field {
  output = List.of(Room)
  args = [
    { name: 'count', field: new InputField('Int'), props: {} }
  ]
  resolve(source: string, args: { count: number }) {
    const rooms: Room[] = [];
    for(let i = 0; i < args.count; i ++) {
      const room = new Room();
      room.name = `My room ${i}`;
      room.tags = [String(i)];
      rooms.push(room)
    }
    return rooms;
  }
}

@defineType(gql`
  type Host {
    name: String!
  }
`)
class Host {
  name: string;

  @mount(new RoomsField())
  rooms: Room[];
}

@defineType(gql`
  type Query {
    host: ${Host}
  }
`)
class Query {
  static host() {
    return new Host();
  }
}

describe('Fields', () => {
  it('should append fields to GraphQLType created', () => {
    const roomObjectType: GraphQLObjectType = getGraphQLType(Room);

    const fields = roomObjectType.getFields();
    expect(fields.tags).not.toBeUndefined();
    expect(fields.tags.type.toString()).toBe('[String]');
  });

  it('should create and execute schema as expected', async () => {
    const schema = new GraphQLSchema({
      query: getGraphQLType(Query),
    });

    expect(printSchema(schema)).toMatchSnapshot();

    const results = await graphql({ schema, source: `
      query {
        host {
          rooms(count: 5) {
            name
            tags
          }
        }
      }
    ` });

    expect(results).toEqual({
      "data": {
        "host": {
          "rooms": [
            { "name": "My room 0",
              "tags": [ "0" ] },
            { "name": "My room 1",
              "tags": [ "1" ] },
            { "name": "My room 2",
              "tags": [ "2" ] },
            { "name": "My room 3",
              "tags": [ "3" ] },
            { "name": "My room 4",
              "tags": [ "4" ] },
          ],
        },
      },
    });
  })
});
