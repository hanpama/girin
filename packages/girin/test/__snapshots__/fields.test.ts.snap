// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fields should create and execute schema as expected 1`] = `
"type Host {
  rooms(count: Int): [Room]
  name: String!
}

type Query {
  host: Host
}

type Room {
  tags: [String]
  name: String!
}
"
`;
